### DISCUSSION TITLE
nvim-java plugin install question

### DISCUSSION THREAD

**Discussion by JohnWickzzz** [ayamir/nvimdots#1101] [Q&A]
最近发现有一个新的java plugin，nvim-java(https://github.com/nvim-java/nvim-java )，在添加插件之后，发现如果是设置了lazy=true，哪怕是添加ft=java之后，依然无法实现java audo dap configuration。如果是设置lazy=false，启动的时候会报错
`Failed to run `config` for nvim-dap

/Users/xx/.config/nvim/lua/modules/utils/init.lua:56: attempt to index field 'colors_name' (a nil value)

# stacktrace:
  - ~/.config/nvim/lua/modules/utils/init.lua:56 _in_ **get_palette**
  - ~/.config/nvim/lua/modules/configs/tool/dap/init.lua:7 _in_ **config**
  - ~/.config/nvim/lua/modules/plugins/custom.lua:64
  - /mason-lspconfig.nvim/lua/mason-lspconfig/init.lua:84 _in_ **fn**
  - /mason.nvim/lua/mason-core/optional.lua:101 _in_ **if_present**
  - /mason-lspconfig.nvim/lua/mason-lspconfig/init.lua:82 _in_ **fn**
  - /mason.nvim/lua/mason-core/functional/list.lua:98 _in_ **each**
  - /mason-lspconfig.nvim/lua/mason-lspconfig/init.lua:92 _in_ **setup_handlers**
  - /mason-lspconfig.nvim/lua/mason-lspconfig/init.lua:50 _in_ **setup**
  - ~/.config/nvim/lua/core/pack.lua:135 _in_ **load_lazy**
  - ~/.config/nvim/lua/core/pack.lua:138
  - ~/.config/nvim/lua/core/init.lua:158 _in_ **load_core**
  - ~/.config/nvim/lua/core/init.lua:180
  - ~/.config/nvim/init.lua:2
  - 请按 ENTER 或其它命令继续`
  
看到报错之后，如果立马按一次enter，就会和之前设置lazy=true类似，不会有java auto dap；
但是如果稍等一会，慢慢按几次enter之后，启动完成，java auto dap configured成功. 
有点困惑为什么会出现这个现象，如果有人使用这个java插件遇到类似的问题，可以探讨一下。 

--------------
Recently I discovered a new neovim plugin nvim-java, after add this plugin, I found if set the lazy=true, even I add ft=java after the lazy=true, the java auto dap configuration can't work. But if set lazy=true, the neovim will occur error after start:
`Failed to run `config` for nvim-dap

/Users/xx/.config/nvim/lua/modules/utils/init.lua:56: attempt to index field 'colors_name' (a nil value)

# stacktrace:
  - ~/.config/nvim/lua/modules/utils/init.lua:56 _in_ **get_palette**
  - ~/.config/nvim/lua/modules/configs/tool/dap/init.lua:7 _in_ **config**
  - ~/.config/nvim/lua/modules/plugins/custom.lua:64
  - /mason-lspconfig.nvim/lua/mason-lspconfig/init.lua:84 _in_ **fn**
  - /mason.nvim/lua/mason-core/optional.lua:101 _in_ **if_present**
  - /mason-lspconfig.nvim/lua/mason-lspconfig/init.lua:82 _in_ **fn**
  - /mason.nvim/lua/mason-core/functional/list.lua:98 _in_ **each**
  - /mason-lspconfig.nvim/lua/mason-lspconfig/init.lua:92 _in_ **setup_handlers**
  - /mason-lspconfig.nvim/lua/mason-lspconfig/init.lua:50 _in_ **setup**
  - ~/.config/nvim/lua/core/pack.lua:135 _in_ **load_lazy**
  - ~/.config/nvim/lua/core/pack.lua:138
  - ~/.config/nvim/lua/core/init.lua:158 _in_ **load_core**
  - ~/.config/nvim/lua/core/init.lua:180
  - ~/.config/nvim/init.lua:2
  - 请按 ENTER 或其它命令继续`

If the error message appears, press enter immediately, will like the lazy=true, java auto dap not work;
but if I wait a while, press enter a couple of times, the java auto dap configured success.
I‘m confused as to why this is happening, if anyone else is experiencing similar issues with this java plugin, could share the info each other.

**Comment by charliie-dev** [comment]
how do u set up nvim-java in your config?

**Comment by JohnWickzzz** [comment]
> how do u set up nvim-java in your config?

I add the nvim-java plugin like this. https://github.com/buptlibo/nvimdots/blob/ed5bd2a69141d7c6e2568d052d2024da6bc310e3/lua/modules/plugins/custom.lua#L37C1-L72C1 

```
custom["nvim-java/nvim-java"] = {
	lazy = true,
	ft = "java",
	event = "LspAttach",
	dependencies = {
		"nvim-java/lua-async-await",
		"nvim-java/nvim-java-core",
		"nvim-java/nvim-java-test",
		"nvim-java/nvim-java-dap",
		"MunifTanjim/nui.nvim",
		"neovim/nvim-lspconfig",
		"mfussenegger/nvim-dap",
		{
			"williamboman/mason.nvim",
			opts = {
				registries = {
					"github:nvim-java/mason-registry",
					"github:mason-org/mason-registry",
				},
			},
		},
		{
			"williamboman/mason-lspconfig.nvim",
			opts = {
				handlers = {
					["jdtls"] = function()
						require("java").setup()
					end,
				},
			},
		},
	},
	opts = {},
}
```

**Reply by Jint-lzxy** [reply]
Wow, this is a huge collection of plugins! I can't guarantee that everything below will 100% work, as I haven't had enough time to fully test this. But I think first off, u don't need to specify dependencies that already exist (see #1087), so the `LazyConfig` for this plugin should be:
```lua
custom["nvim-java/nvim-java"] = {
	lazy = true,
	ft = "java",
	dependencies = {
		"nvim-java/lua-async-await",
		"nvim-java/nvim-java-core",
		"nvim-java/nvim-java-test",
		"nvim-java/nvim-java-dap",
		"MunifTanjim/nui.nvim",
	},
}
```
Also, for `mason.nvim` and `mason-lspconfig.nvim`, you need to modify their configs individually, like this:
1. Append
```lua
registries = {
	"github:nvim-java/mason-registry",
	"github:mason-org/mason-registry",
},
```
after [this line](https://github.com/ayamir/nvimdots/blob/8bcc4099aa17812abfcf35a9dea0e5673d461bb5/lua/modules/configs/completion/mason.lua#L32) (or use user configs, see [Wiki: Usage#steps-to-modify-plugin-configs](https://github.com/ayamir/nvimdots/wiki/Usage#steps-to-modify-plugin-configs))

2. Create `jdtls.lua` under `lua/modules/configs/completion/servers` (or equivalent user config directory), and fill this file with the following content:
```lua
return function()
	require("java").setup()
end
```

Finally, add this line [here](https://github.com/ayamir/nvimdots/blob/8bcc4099aa17812abfcf35a9dea0e5673d461bb5/lua/modules/configs/tool/dap/init.lua#L78):
```lua
require("java").dap.config_dap()
```

**Comment by JohnWickzzz** [comment]
> Wow, this is a huge collection of plugins! I can't guarantee that everything below will 100% work, as I haven't had enough time to fully test this. But I think first off, u don't need to specify dependencies that already exist (see #1087), so the `LazyConfig` for this plugin should be:
> 
> ```lua
> custom["nvim-java/nvim-java"] = {
> 	lazy = true,
> 	ft = "java",
> 	dependencies = {
> 		"nvim-java/lua-async-await",
> 		"nvim-java/nvim-java-core",
> 		"nvim-java/nvim-java-test",
> 		"nvim-java/nvim-java-dap",
> 		"MunifTanjim/nui.nvim",
> 	},
> }
> ```
> 
> Also, for `mason.nvim` and `mason-lspconfig.nvim`, you need to modify their configs individually, like this:
> 
> 1. Append
> 
> ```lua
> registries = {
> 	"github:nvim-java/mason-registry",
> 	"github:mason-org/mason-registry",
> },
> ```
> 
> after [this line](https://github.com/ayamir/nvimdots/blob/8bcc4099aa17812abfcf35a9dea0e5673d461bb5/lua/modules/configs/completion/mason.lua#L32) (or use user configs, see [Wiki: Usage#steps-to-modify-plugin-configs](https://github.com/ayamir/nvimdots/wiki/Usage#steps-to-modify-plugin-configs))
> 
> 2. Create `jdtls.lua` under `lua/modules/configs/completion/servers` (or equivalent user config directory), and fill this file with the following content:
> 
> ```lua
> return function()
> 	require("java").setup()
> end
> ```
> 
> Finally, add this line [here](https://github.com/ayamir/nvimdots/blob/8bcc4099aa17812abfcf35a9dea0e5673d461bb5/lua/modules/configs/tool/dap/init.lua#L78):
> 
> ```lua
> require("java").dap.config_dap()
> ```


Oh, after update config. Even if I only open the neovim, a lot of error message occurs like this:

```
could not find an active jdtls client
stack traceback:
	...are/nvim/site/lazy/nvim-java/lua/java/handlers/error.lua:18: in function <...are/nvim/site/lazy/nvim-java/lua/java/handlers/error.lua:17>
	.../share/nvim/site/lazy/nvim-java/lua/java/utils/jdtls.lua:9: in function 'jdtls'
	...ocal/share/nvim/site/lazy/nvim-java/lua/java/api/dap.lua:28: in function <...ocal/share/nvim/site/lazy/nvim-java/lua/java/api/dap.lua:26>


dap configuration failed
...are/nvim/site/lazy/nvim-java/lua/java/handlers/error.lua:28: 
could not find an active jdtls client
stack traceback:
	...are/nvim/site/lazy/nvim-java/lua/java/handlers/error.lua:18: in function <...are/nvim/site/lazy/nvim-java/lua/java/handlers/error.lua:17>
	.../share/nvim/site/lazy/nvim-java/lua/java/utils/jdtls.lua:9: in function 'jdtls'
	...ocal/share/nvim/site/lazy/nvim-java/lua/java/api/dap.lua:28: in function <...ocal/share/nvim/site/lazy/nvim-java/lua/java/api/dap.lua:26>
stack traceback:
	...are/nvim/site/lazy/nvim-java/lua/java/handlers/error.lua:18: in function 'error_handler'
	...m/site/lazy/nvim-java-core/lua/java-core/utils/async.lua:32: in function 'step'
	...m/site/lazy/nvim-java-core/lua/java-core/utils/async.lua:54: in function 'config_dap'
	.../brucelee/.config/nvim/lua/modules/configs/tool/dap/init.lua:79: in function 'config'
	.../share/nvim/site/lazy/lazy.nvim/lua/lazy/core/loader.lua:362: in function <.../share/nvim/site/lazy/lazy.nvim/lua/lazy/core/loader.lua:360>
	[C]: in function 'xpcall'
	...al/share/nvim/site/lazy/lazy.nvim/lua/lazy/core/util.lua:113: in function 'try'
	.../share/nvim/site/lazy/lazy.nvim/lua/lazy/core/loader.lua:377: in function 'config'
	.../share/nvim/site/lazy/lazy.nvim/lua/lazy/core/loader.lua:344: in function '_load'
	.../share/nvim/site/lazy/lazy.nvim/lua/lazy/core/loader.lua:187: in function 'load'
	.../share/nvim/site/lazy/lazy.nvim/lua/lazy/core/loader.lua:514: in function 'auto_load'
	...
	...e/lazy/mason-lspconfig.nvim/lua/mason-lspconfig/init.lua:92: in function 'setup_handlers'
	.../nvim/lua/modules/configs/completion/mason-lspconfig.lua:60: in function 'setup'
	...brucelee/.config/nvim/lua/modules/configs/completion/lsp.lua:4: in function 'config'
	.../share/nvim/site/lazy/lazy.nvim/lua/lazy/core/loader.lua:362: in function <.../share/nvim/site/lazy/lazy.nvim/lua/lazy/core/loader.lua:360>
	[C]: in function 'xpcall'
	...al/share/nvim/site/lazy/lazy.nvim/lua/lazy/core/util.lua:113: in function 'try'
	.../share/nvim/site/lazy/lazy.nvim/lua/lazy/core/loader.lua:377: in function 'config'
	.../share/nvim/site/lazy/lazy.nvim/lua/lazy/core/loader.lua:344: in function '_load'
	.../share/nvim/site/lazy/lazy.nvim/lua/lazy/core/loader.lua:187: in function 'load'
	...nvim/site/lazy/lazy.nvim/lua/lazy/core/handler/event.lua:83: in function <...nvim/site/lazy/lazy.nvim/lua/lazy/core/handler/event.lua:72>



Failed to run `config` for nvim-dap

...are/nvim/site/lazy/nvim-java/lua/java/handlers/error.lua:28: 
dap configuration failed
...are/nvim/site/lazy/nvim-java/lua/java/handlers/error.lua:28: 
could not find an active jdtls client
stack traceback:
	...are/nvim/site/lazy/nvim-java/lua/java/handlers/error.lua:18: in function <...are/nvim/site/lazy/nvim-java/lua/java/handlers/error.lua:17>
	.../share/nvim/site/lazy/nvim-java/lua/java/utils/jdtls.lua:9: in function 'jdtls'
	...ocal/share/nvim/site/lazy/nvim-java/lua/java/api/dap.lua:28: in function <...ocal/share/nvim/site/lazy/nvim-java/lua/java/api/dap.lua:26>
stack traceback:
	...are/nvim/site/lazy/nvim-java/lua/java/handlers/error.lua:18: in function 'error_handler'
	...m/site/lazy/nvim-java-core/lua/java-core/utils/async.lua:32: in function 'step'
	...m/site/lazy/nvim-java-core/lua/java-core/utils/async.lua:54: in function 'config_dap'
	.../brucelee/.config/nvim/lua/modules/configs/tool/dap/init.lua:79: in function 'config'
	.../share/nvim/site/lazy/lazy.nvim/lua/lazy/core/loader.lua:362: in function <.../share/nvim/site/lazy/lazy.nvim/lua/lazy/core/loader.lua:360>
	[C]: in function 'xpcall'
	...al/share/nvim/site/lazy/lazy.nvim/lua/lazy/core/util.lua:113: in function 'try'
	.../share/nvim/site/lazy/lazy.nvim/lua/lazy/core/loader.lua:377: in function 'config'
	.../share/nvim/site/lazy/lazy.nvim/lua/lazy/core/loader.lua:344: in function '_load'
	.../share/nvim/site/lazy/lazy.nvim/lua/lazy/core/loader.lua:187: in function 'load'
	.../share/nvim/site/lazy/lazy.nvim/lua/lazy/core/loader.lua:514: in function 'auto_load'
	...
	...e/lazy/mason-lspconfig.nvim/lua/mason-lspconfig/init.lua:92: in function 'setup_handlers'
	.../nvim/lua/modules/configs/completion/mason-lspconfig.lua:60: in function 'setup'
	...brucelee/.config/nvim/lua/modules/configs/completion/lsp.lua:4: in function 'config'
	.../share/nvim/site/lazy/lazy.nvim/lua/lazy/core/loader.lua:362: in function <.../share/nvim/site/lazy/lazy.nvim/lua/lazy/core/loader.lua:360>
	[C]: in function 'xpcall'
	...al/share/nvim/site/lazy/lazy.nvim/lua/lazy/core/util.lua:113: in function 'try'
	.../share/nvim/site/lazy/lazy.nvim/lua/lazy/core/loader.lua:377: in function 'config'
	.../share/nvim/site/lazy/lazy.nvim/lua/lazy/core/loader.lua:344: in function '_load'
	.../share/nvim/site/lazy/lazy.nvim/lua/lazy/core/loader.lua:187: in function 'load'
	...nvim/site/lazy/lazy.nvim/lua/lazy/core/handler/event.lua:83: in function <...nvim/site/lazy/lazy.nvim/lua/lazy/core/handler/event.lua:72>

# stacktrace:
  - ~/.config/nvim/lua/modules/configs/tool/dap/init.lua:79 _in_ **config**
  - ~/.config/nvim/lua/modules/configs/completion/servers/jdtls.lua:2 _in_ **custom_handler**
  - ~/.config/nvim/lua/modules/configs/completion/mason-lspconfig.lua:44
  - /mason-lspconfig.nvim/lua/mason-lspconfig/init.lua:84 _in_ **fn**
  - /mason.nvim/lua/mason-core/optional.lua:101 _in_ **if_present**
  - /mason-lspconfig.nvim/lua/mason-lspconfig/init.lua:82 _in_ **fn**
  - /mason.nvim/lua/mason-core/functional/list.lua:98 _in_ **each**
  - /mason-lspconfig.nvim/lua/mason-lspconfig/init.lua:92 _in_ **setup_handlers**
  - ~/.config/nvim/lua/modules/configs/completion/mason-lspconfig.lua:60 _in_ **setup**
  - ~/.config/nvim/lua/modules/configs/completion/lsp.lua:4 _in_ **config**
```

But if I open a java file, neovim start normal, but the jdtls servers doesn't work. So If I want to debug project, it will also show the dap error.

**Reply by Jint-lzxy** [reply]
> could not find an active jdtls client

I just checked out the source code of `nvim-java`. Looks like this plugin still needs the user to explicitly register `jdtls` in `nvim-lspconfig` for it to work properly. It also doesn't require a separate initialization (loop). So its `LazyConfig` should be like this:
```lua

custom["nvim-java/nvim-java"] = {
	lazy = true,
	-- ft = "java",
	dependencies = {
		"nvim-java/lua-async-await",
		"nvim-java/nvim-java-core",
		"nvim-java/nvim-java-test",
		"nvim-java/nvim-java-dap",
		"MunifTanjim/nui.nvim",
	},
}
```
and `jdtls.lua` should be filled with the following content:
```lua
require("java").setup()

-- https://github.com/neovim/nvim-lspconfig/blob/master/lua/lspconfig/server_configurations/jdtls.lua
return {}
```

> Failed to run `config` for nvim-dap

I think this is more or less a plugin-side issue, as `nvim-java-dap` doesn't support lazy loading. u could try specifying `nvim-java/nvim-java-dap` as a dependency for `mfussenegger/nvim-dap` and set `lazy = true` for it, or change the `LazyConfig` of `mfussenegger/nvim-dap` to `lazy = false` and [set a very low priority](https://github.com/folke/lazy.nvim#-plugin-spec) for it (like `10`).

But in the case that all of these methods don't work, I suggest u (1) ask the author of this plugin if he has any plan to support lazy-loaded dap config (i.e., only initialize the debugging configs **after** the user explicitly calls `require("java").dap.config_dap()`), or (2) just set the specs for all related plugins to `lazy = false` (but this might lead to overly long startup time).

**Reply by JohnWickzzz** [reply]
Thanks for your help! 👍 I will try it.
